When converting HTML that contains an attachment to an attributed string and back using `HTMLConverter::editingAttributedString`, the attachment is lost

rr-codes opened the pull request:
#### af90843c7e73e53a874763252a77875ab196857c
<pre>
When converting HTML that contains an attachment to an attributed string and back using `HTMLConverter::editingAttributedString`, the attachment is lost
<a href="https://bugs.webkit.org/show_bug.cgi?id=274948">https://bugs.webkit.org/show_bug.cgi?id=274948</a>
<a href="https://rdar.apple.com/120873087">rdar://120873087</a>

Reviewed by Wenson Hsieh.

Add support for non-image attachments in `HTMLConverter::editingAttributedString`. To do this,
encode the attachment id into an NSTextAttachment, and then decode it back to the existing HTMLAttachmentElement
element from within `WebContentReaderCocoa`.

Add a TextIterator option to maintain consistency between the character count produced by TextIterator
and the length of the string produced by `HTMLConverter::editingAttributedString`.

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::handleReplacedElement):
* Source/WebCore/editing/TextIteratorBehavior.h:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/cocoa/HTMLConverter.h:
(WebCore::editingAttributedString):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(fileWrapperForElement):
(attributedStringWithAttachmentForElement):
(WebCore::editingAttributedString):
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::replaceRichContentWithAttachments):
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::dictionaryPopupInfoForRange):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::attributedSubstringForCharacterRangeAsync):
* Source/WebKitLegacy/mac/WebView/WebImmediateActionController.mm:
(+[WebImmediateActionController _dictionaryPopupInfoForRange:inFrame:indicatorOptions:transition:]):

Canonical link: <a href="https://commits.webkit.org/279605@main">https://commits.webkit.org/279605@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9e40e343f18aa730429f2a563a96e082eed529d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53960 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57238 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4686 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56263 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40841 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4571 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3092 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56057 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46680 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24832 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4002 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2836 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4204 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58831 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/51106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30327 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46817 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11752 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7989 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2393,7 +2391,23 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
+    if (!data)
+        return nil;
+
+    // Use a filename prefixed with a sentinal value to indicate that the data is corresponding
```
```suggestion
    // Use a filename prefixed with a sentinel value to indicate that the data is corresponding
```

whsieh commented on Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
```diff
@@ -379,6 +379,16 @@ static void replaceRichContentWithAttachments(LocalFrame& frame, DocumentFragmen
         if (!parent)
             continue;
 
+        // If the filename begins with this sentinal value, this means that an existing attachment should be used.
```
```suggestion
        // If the filename begins with this sentinel value, this means that an existing attachment should be used.
```

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -49,24 +49,26 @@
 
 // MARK: Static utility helper methods.
 
+static constexpr auto defaultTextIteratorBehaviors = WebCore::TextIteratorBehaviors { WebCore::TextIteratorBehavior::EmitsObjectReplacementCharactersForImagesOnly, WebCore::TextIteratorBehavior::EmitsObjectReplacementCharactersForAttachmentsOnly };
```
It seems a bit weird to have values of both `ImagesOnly` and `AttachmentsOnly`. Does that mean only image-backed attachments emit replacement characters?

If not, then it might be clearer to drop the `-Only` part.

Nit - I think it be good to guard this with `ENABLE(ATTACHMENT_ELEMENT)` as well, even though attachment elements are effectively always enabled for Cocoa platforms, just so that the compile-time guards are consistent between where the enum value is used, and where it's declared.

whsieh commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2393,7 +2391,23 @@ static String preferredFilenameForElement(const HTMLImageElement& element)
+    // to an existing HTMLAttachmentElement.
+
+    RetainPtr wrapper = adoptNS([[NSFileWrapper alloc] initRegularFileWithContents:data.get()]);
+    [wrapper setPreferredFilename:makeString(WebContentReader::placeholderAttachmentFilenamePrefix, identifier)];
```
This is...pretty hacky, but I'm not sure I have a better suggestion immediately. I suppose it's okay for now, as long as a maliciously-crafted email can't cause, say, a crash in either the web process or Mail by purposely leaving an attachment named `webkit-attachment-XXXXX`.

Does it work if you make this a separate (custom) key in the `NSAttributedString`, that maps to the attachment ID?

rr-codes replied:
> Does it work if you make this a separate (custom) key in the NSAttributedString, that maps to the attachment

It does not, since NSHtMLWriter just ends up dropping it

rr-codes committed:
When converting HTML that contains an attachment to an attributed string and back using `HTMLConverter::editingAttributedString`, the attachment is lost
https://bugs.webkit.org/show_bug.cgi?id=274948
rdar://120873087

Reviewed by Wenson Hsieh.

Add support for non-image attachments in `HTMLConverter::editingAttributedString`. To do this,
encode the attachment id into an NSTextAttachment, and then decode it back to the existing HTMLAttachmentElement
element from within `WebContentReaderCocoa`.

Add a TextIterator option to maintain consistency between the character count produced by TextIterator
and the length of the string produced by `HTMLConverter::editingAttributedString`.

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::handleReplacedElement):
* Source/WebCore/editing/TextIteratorBehavior.h:
* Source/WebCore/editing/WebContentReader.h:
* Source/WebCore/editing/cocoa/HTMLConverter.h:
(WebCore::editingAttributedString):
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(fileWrapperForElement):
(attributedStringWithAttachmentForElement):
(WebCore::editingAttributedString):
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::replaceRichContentWithAttachments):
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::dictionaryPopupInfoForRange):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::attributedSubstringForCharacterRangeAsync):
* Source/WebKitLegacy/mac/WebView/WebImmediateActionController.mm:
(+[WebImmediateActionController _dictionaryPopupInfoForRange:inFrame:indicatorOptions:transition:]):

Canonical link: https://commits.webkit.org/279605@main
